CSP 1308 - SynchroniseApprenticeDetailsFunction

shaun-downey-education opened the pull request:
- Introduced a new Azure function to synchronize apprentice details between AAN.Hub and the ApprenticeAccounts databases.
- Additional updates added to manage RestEase api integration.

shaun-downey-education committed:
CSP1308: Initial Merge into new branch

shaun-downey-education committed:
CSP1308: Initial Merge into new branch

shaun-downey-education committed:
CSP-1308: Addition of SynchroniseApprenticeDetailsFunction with included unit tests

shaun-downey-education committed:
CSP-1308: Addition of SynchroniseApprenticeDetailsFunction with included unit tests

shaun-downey-education committed:
SCP-1308: Removed unused configuration property

shaun-downey-education committed:
Merge branch 'CSP-1308_Apprentice_Sync_Daily_Job' of https://github.com/SkillsFundingAgency/das-aan-hub-jobs into CSP-1308_Apprentice_Sync_Daily_Job

shaun-downey-education committed:
CSP-1308: Replace IHTTPClientFactory with RestEase

shaun-downey-education committed:
CSP-1380: Replaced IHTTPClientFactory with RestEase.
Added InnerApi configuration for the Apprentice Api.
Updated unit tests to the new Restease client.

shaun-downey-education committed:
CSP-1308: Additional unit test coverage for SynchroniseApprentices.

shaun-downey-education committed:
CSP-1308: Sonar Cloud code Inconsistency updates.

shaun-downey-education committed:
CSP-1308: Sonar Cloud code inonsistency (2)

shaun-downey-education committed:
CSP-1308: Sonar Cloud code maintainability updates.

shaun-downey-education committed:
CSP-1308: Sonar Cloud code maintainability.

DevenKShah commented on src/SFA.DAS.AAN.Hub.Data/Repositories/JobAuditRepository.cs:
```diff
@@ -0,0 +1,50 @@
+
+        public async Task<JobAudit?> GetMostRecentJobAudit(CancellationToken cancellationToken)
+        {
+            try
```
I dont like suppressing errors, we are not expecting this operations to fail and hence we should not be catching any errors, this will allow the actual exceptions to bubble up and Application Insights can then show us exactly where it occurred. Please remove all the instances, across files, where try catches are suppressing exceptions.

DevenKShah commented on src/SFA.DAS.AAN.Hub.Data/Repositories/ApprenticeRepository.cs:
Remove try catches

DevenKShah commented on src/SFA.DAS.AAN.Hub.Jobs.UnitTests/FixtureSpecimens/GuidSpecimenBuilder.cs:
```diff
@@ -0,0 +1,20 @@
+namespace SFA.DAS.AAN.Hub.Jobs.UnitTests.FixtureSpecimens
+{
+    [ExcludeFromCodeCoverage]
+    public class GuidSpecimenBuilder : ISpecimenBuilder
```
Why do we need this?

DevenKShah commented on src/SFA.DAS.AAN.Hub.Jobs.UnitTests/SFA.DAS.AAN.Hub.Jobs.UnitTests.csproj:
```diff
@@ -32,4 +34,7 @@
     <ProjectReference Include="..\SFA.DAS.AAN.Hub.Jobs\SFA.DAS.AAN.Hub.Jobs.csproj" />
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="Api\" />
```
Empty folder?

DevenKShah commented on src/SFA.DAS.AAN.Hub.Jobs/Api/Response/ApprenticeSyncDto.cs:
```diff
@@ -0,0 +1,28 @@
+    [ExcludeFromCodeCoverage]
+    public class ApprenticeSyncDto
+    {
+        [JsonProperty("apprenticeId")]
```
These attributes are not adding any value as this is the default behaviour.

DevenKShah commented on src/SFA.DAS.AAN.Hub.Jobs/Services/SynchroniseApprenticeDetailsService.cs:
```diff
@@ -0,0 +1,154 @@
+            _logger.LogError(_exception, "{MethodName} Failed.", nameof(SynchroniseApprentices));
+            await RecordAudit(audit, null, cancellationToken);
+            return default;
+        }
```
I think we decided to not to record any failures in audit. I suggest remove this try catch completely.

DevenKShah commented on src/SFA.DAS.AAN.Hub.Jobs/Services/SynchroniseApprenticeDetailsService.cs:
```diff
@@ -0,0 +1,154 @@
+
+            var response = await QueryApprenticeApi(apprentices, cancellationToken);
+
+            if(response is null)
```
Response will never be null, you can check http status code is not success, but in this case we do not want to create audit record

DevenKShah commented on src/SFA.DAS.AAN.Hub.Data/Repositories/JobAuditRepository.cs:
```diff
@@ -0,0 +1,50 @@
+            try
+            {
+                _context.JobAudits.Add(jobAudit);
+                await _context.SaveChangesAsync(cancellationToken);
```
Save changes should not be added to the repository as it will be non-transactional commit. Remove save changes from the repository and move it into service. So all the changes are save under one transaction. FYI the dbcontext is scoped in DI so an intrinsic transaction is created for you.

shaun-downey-education committed:
CSP-1308: Rewrite on SynchroniseApprenticeDetailsFunction for:
- Accessing members through members directly.
- Writing function in a transctional pattern.

shaun-downey-education committed:
- CSP-1308: Sonarcloud code recommendations.

shaun-downey-education committed:
- CSP-1308: Remove appSettings.

DevenKShah commented on src/SFA.DAS.AAN.Hub.Data/Interfaces/IMemberRepository.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface IMemberRepository
+    {
+        Task<List<Member>> GetActiveMembers(CancellationToken cancellationToken);
```
GetActiveApprenticeMembers

shaun-downey-education committed:
CSP-1308: Renamed GetActiveMembers to GetActiveApprenticeMembers. Code Review Comments.

DevenKShah commented on src/SFA.DAS.AAN.Hub.Data/Repositories/MemberRepository.cs:
```diff
@@ -0,0 +1,23 @@
+
+        public async Task<List<Member>> GetActiveApprenticeMembers(CancellationToken cancellationToken)
+        {
+            return await _context.Members.Where(m => m.Email != m.Id.ToString())
```
Need to add UserType == Apprentice here

DevenKShah commented on src/SFA.DAS.AAN.Hub.Jobs/Configuration/ApplicationConfiguration.cs:
```diff
@@ -6,6 +6,7 @@ public class ApplicationConfiguration
 {
     public Uri ApprenticeAanBaseUrl { get; set; }
     public Uri EmployerAanBaseUrl { get; set; }
+    public Uri ApprenticeAccountsApiBaseUrl { get; set; }
```
I was expecting a property of type `InnerApiConfiguration` here with name `ApprenticeAccountsApiConfiguration` which matches the config name.

shaun-downey-education committed:
CSP-1308: Bind ApprenticeAccountsApiConfiguration directly to ApplicationConfiguration and resolve RestEase AddApprenticeAccountsApiClient settings from ApprenticeAccountsApiConfiguration directly.

shaun-downey-education committed:
CSP-1308: Remove ununsed GetApiConfiguration method.

shaun-downey-education committed:
CSP-1308: Filter apprentice members by UserType = Apprentice.

shaun-downey-education committed:
CSP-1308: Introduced a new integration test project.

shaun-downey-education committed:
CSP-1308: Upgrade tranistive packages - IntegrationTests solution.

shaun-downey-education committed:
CSP-1308: Integration unit test project source change.

shaun-downey-education committed:
CSP-1308: Updated Integration test project name to match that of the build pipeline naming convention.

DevenKShah committed:
CSP-1308 Added function schedule to environment variable

DevenKShah committed:
CSP-1308 Fixed package warnings

shaun-downey-education committed:
CSP-1308: Update GetMostRecentJobAudit method to do a generic comparison of strings.

DevenKShah committed:
CSP-1308 Downgrade JWT packages to get around auth error

DevenKShah committed:
CSP-1308 Added app role assignments step in deployment

Commit:
Added in param

DevenKShah committed:
CSP-1308 Fixed issue in deploy template

Commit:
Added something else in yaml

shaun-downey-education committed:
CSP-1308: Updated synchronise members function to return a tuple explaining the number of both processed & updated members.

shaun-downey-education committed:
CSP-1308: Fixed grammar issue.
